### DISCUSSION TITLE
[Bug] Vertx worker thread can't get TraceId

### DISCUSSION THREAD

**Discussion by naah69** [apache/skywalking#10190] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

In vertx 4.x agent,`SWVertxTracer#toEntrySpan` parse context in eventloop thread.
 ContextManager can't be used in worker thread.
<img width="1508" alt="image" src="https://user-images.githubusercontent.com/25682169/208832293-d23d1673-f288-4108-a5b1-d08c66e28426.png">
<img width="1562" alt="image" src="https://user-images.githubusercontent.com/25682169/208833274-0979040a-4ded-47cf-9f44-d66a6a882aea.png">

### What you expected to happen

TraceId is N/A

### How to reproduce

use vertx agent and invoke  `ContextManager#getTraceId` in worker thread.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by naah69** [comment]
i had used filter to deal it.
but i think it is so ugly.
is there any better way to reslove it ?
<img width="746" alt="image" src="https://user-images.githubusercontent.com/25682169/208879417-1727a81d-6025-45db-8ab8-46a0216b5fbf.png">

**Comment by wu-sheng** [comment]
Let's see @BFergerson has any suggestions.

@naah69 For a community, there is nothing urgent. All PMC members are volunteers here, they just contribute and answer questions for their own interests.

**Comment by BFergerson** [comment]
I gave it a look. It might be an issue but it's a bit hard to tell without a reproducible example. It may not be an issue with worker threads but just how the Vertx plugin uses async spans. Calling `ContextManager.getGlobalTraceId()` in a `RoutingContext` handler returns `N/A` regardless if it's the event loop or worker thread. The tracing data is stored transparently in the request context.

Example:
![image](https://user-images.githubusercontent.com/3278877/208890419-3f71c64f-7bc2-4d1b-beb4-5535f7b096b3.png)

**Reply by naah69** [reply]
yes，maybe we should find some invocation point to intercept to continue context snapshot in routing context.

**Reply by BFergerson** [reply]
I don't think I've considered how that would work before.

@wu-sheng, have you run into this situation before with the async spans? For example:
- an entry async span starts
- the user adds `@Trace` (or equivalent) to start a local span
- the entry async span stops

I haven't tested it but my gut feeling is the local span won't be assigned to the entry span as it would not be the active span.

---

I think the Vert.x plugin could add local trace spans before it gives back control to user code to fix this issue but I'm curious how other async span plugins handle it.

**Reply by wu-sheng** [reply]
From the agent core perspective, once the span is not finished, no matter in or async mode, the next span could be in the same trace context and treat the span as the parent.

**Reply by wu-sheng** [reply]
But if it has done stop(not async stop), it is gone.

**Reply by BFergerson** [reply]
Yeah, I verified that's the case. I don't think there is an issue here then.

@naah69, what exactly are you trying to achieve?
